### PR TITLE
swift: 5.1.1 -> 5.4.2

### LINKED ISSUE #97859: Update Swift to 5.3
## Issue description

I'm trying to work on a project using the [Vapor framework][vapor], whose stable versions require at least Swift 5.2. The current version in nixpkgs [is still 5.1][one]. I've tried to do the legwork of upgrading Swift in nixpkgs myself, but I'm not familiar enough with the Swift codebase and have quickly found myself in over my head!

In any case, I'd like to be able to use the current version of Swift without resorting to their docker images, and I'm happy to help out any way I can.

### Steps to reproduce

N/A

## Technical details

`nix-shell -p nix-info --run "nix-info -m"`:

 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.61, NixOS, 20.03.2862.925ae0dee63 (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.6`
 - channels(root): `"nixos-20.03.2862.925ae0dee63, home-manager-20.03, nixos-hardware"`
 - channels(mistress): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

[vapor]: https://github.com/vapor/vapor
[one]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/swift/default.nix#L38

### PR DESCRIPTION
**Pull request by trepetti**
###### Motivation for this change

Fix Swift build, remove CVE-2020-9861, update Swift three minor versions. Fixes #97859. We can update from here to enable building Swift for Darwin based on #37473 and https://github.com/stephank/nixpkgs/commit/2b035eff96722093ac458fe34fcef31ea304c7ae if there is still interest. My set up is Linux-only for now, though.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by trepetti** [comment]
Force push to switch from swift-check-all to swift-check-validation to avoid ofborg timeout.

File: pkgs/development/compilers/swift/default.nix [hunk 1]
```diff
@@ -35,73 +36,102 @@
...
+      repo = "swift-llbuild";
+      sha256 = "0d7sj5a9b5c1ry2209cpccic5radf9s48sp1lahqzmd1pdx3n7pi";
     };
-    cmark = fetch {
-      repo = "swift-cmark";
-      sha256 = "079smm79hbwr06bvghd2sb86b8gpkprnzlyj9kh95jy38xhlhdnj";
+    argumentParser = fetchFromGitHub {
+      owner = "apple";
+      repo = "swift-argument-parser";
+      rev = "${swiftArgumentParserVersion}";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/swift/default.nix:71]:
```suggestion
      rev = swiftArgumentParserVersion;
```

File: pkgs/development/compilers/swift/default.nix [hunk 2]
```diff
@@ -35,73 +36,102 @@
...
-      repo = "swift";
-      sha256 = "0m4r1gzrnn0s1c7haqq9dlmvpqxbgbkbdfmq6qaph869wcmvdkvy";
+    # TODO: possibly re-add integration-tests.
+    # Linux does not support Xcode playgrounds.
+    # We provide our own ninja.
+    # We provider our own icu.
+    yams = fetchFromGitHub {
+      owner = "jpsim";
+      repo = "Yams";
+      rev = "${yamsVersion}";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/swift/default.nix:112]:
```suggestion
      rev = yamsVersion;
```

File: pkgs/development/compilers/swift/default.nix [hunk 3]
```diff
@@ -35,73 +36,102 @@
...
+      sha256 = "1ap3hiq2jd3cn10d8d674xysq27by878mvq087a80681r8cdivn3";
+    };
+    sourcekitLsp = fetchSwiftRelease {
+      repo = "sourcekit-lsp";
+      sha256 = "02m9va0lsn2hnwkmgrbgj452sbyaswwmq14lqvxgnb7gssajv4gc";
+    };
+    format = fetchFromGitHub {
+      owner = "apple";
+      repo = "swift-format";
+      rev = "${swiftFormatVersion}";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/swift/default.nix:128]:
```suggestion
      rev = swiftFormatVersion;
```

File: pkgs/development/compilers/swift/default.nix [hunk 4]
```diff
@@ -136,24 +167,26 @@ stdenv.mkDerivation {
...
     findutils
+    git
     gnumake
     libtool
     makeWrapper
     ninja
     perl
     pkg-config
-    python
+    (python3.withPackages (ps: [ps.six]))
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/swift/default.nix:177]: Please define python once in a let in and reuse it over the file.

File: pkgs/development/compilers/swift/default.nix [hunk 5]
```diff
@@ -164,91 +197,89 @@ stdenv.mkDerivation {
...
+     substituteInPlace llvm-project/clang/lib/Driver/ToolChains/Linux.cpp \
+       --replace 'SysRoot + "/lib' '"${glibc}/lib" "'
+     substituteInPlace llvm-project/clang/lib/Driver/ToolChains/Linux.cpp \
+       --replace 'SysRoot + "/usr/lib' '"${glibc}/lib" "'
+     substituteInPlace llvm-project/clang/lib/Driver/ToolChains/Linux.cpp \
+       --replace 'LibDir = "lib";' 'LibDir = "${glibc}/lib";'
+     substituteInPlace llvm-project/clang/lib/Driver/ToolChains/Linux.cpp \
+       --replace 'LibDir = "lib64";' 'LibDir = "${glibc}/lib";'
+     substituteInPlace llvm-project/clang/lib/Driver/ToolChains/Linux.cpp \
+       --replace 'LibDir = X32 ? "libx32" : "lib64";' 'LibDir = "${glibc}/lib";'
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/swift/default.nix:262]:
```suggestion
     substituteInPlace llvm-project/clang/lib/Driver/ToolChains/Linux.cpp \
       --replace 'SysRoot + "/lib' '"${glibc}/lib" "' \
       --replace 'SysRoot + "/usr/lib' '"${glibc}/lib" "' \
       --replace 'LibDir = "lib";' 'LibDir = "${glibc}/lib";' \
       --replace 'LibDir = "lib64";' 'LibDir = "${glibc}/lib";' \
       --replace 'LibDir = X32 ? "libx32" : "lib64";' 'LibDir = "${glibc}/lib";'
```

File: pkgs/development/compilers/swift/default.nix (hunk 1 above)
Comment (trepetti) [reply pkgs/development/compilers/swift/default.nix]: Thank you!

File: pkgs/development/compilers/swift/default.nix (hunk 5 above)
Comment (trepetti) [reply pkgs/development/compilers/swift/default.nix]: Fixed.

**Comment by trepetti** [comment]
Running `nixpkgs-review pr --post-result 132325`. It will take about 3-4 hours on my machine. Unfortunately, the ofborg tests will timeout and I don't have any AArch64 hardware to use.

**Comment by taku0** [comment]
It is working.

```
[nix-shell:~/.cache/nixpkgs-review/pr-132325]$ ./results/swift/bin/swift
Welcome to Swift version 5.4.2 (swift-5.4.2-RELEASE).
Type :help for assistance.
  1> import Foundation
  2> FileManager.`default`.contentsOfDirectory(atPath: "/")
$R0: [String] = 17 values {
  [0] = "root"
  [1] = "boot"
  [2] = "var"
  [3] = "home"
  [4] = "tmp"
  [5] = "proc"
  [6] = "usr"
  [7] = "etc"
  [8] = "srv"
  [9] = "bin"
  [10] = "dev"
  [11] = "opt"
  [12] = "lost+found"
  [13] = "media"
  [14] = "sys"
  [15] = "run"
  [16] = "nix"
}
```

**Comment by taku0** [comment]
Could not build [Vapor](https://docs.vapor.codes/4.0/install/linux/). I'm not sure this is either a problem of Vapor or Swift.

<details><summary>log</summary>

```
[nix-shell:~/opt]$ git clone --depth 1 https://github.com/vapor/toolbox.git        
Cloning into 'toolbox'...
remote: Enumerating objects: 52, done.
remote: Counting objects: 100% (52/52), done.
remote: Compressing objects: 100% (48/48), done.
remote: Total 52 (delta 1), reused 22 (delta 0), pack-reused 0
Receiving objects: 100% (52/52), 26.31 KiB | 8.77 MiB/s, done.
Resolving deltas: 100% (1/1), done.

[nix-shell:~/opt]$ cd toolbox/

[nix-shell:~/opt/toolbox]$ LC_ALL=C make
swiftc ./scripts/build.swift
./scripts/build.swift:48:10: warning: 'launchPath' is deprecated: renamed to 'executableURL'
    task.launchPath = "/usr/bin/env"
         ^
./scripts/build.swift:48:10: note: use 'executableURL' instead
    task.launchPath = "/usr/bin/env"
         ^~~~~~~~~~
         executableURL
./scripts/build.swift:50:10: warning: 'launch()' is deprecated: renamed to 'run'
    task.launch()
         ^
./scripts/build.swift:50:10: note: use 'run' instead
    task.launch()
         ^~~~~~
         run
./scripts/build.swift:61:10: warning: 'launchPath' is deprecated: renamed to 'executableURL'
    task.launchPath = "/usr/bin/env"
         ^
./scripts/build.swift:61:10: note: use 'executableURL' instead
    task.launchPath = "/usr/bin/env"
         ^~~~~~~~~~
         executableURL
./scripts/build.swift:69:10: warning: 'launch()' is deprecated: renamed to 'run'
    task.launch()
         ^
./scripts/build.swift:69:10: note: use 'run' instead
    task.launch()
         ^~~~~~
         run
/nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
./build
$ swift build --disable-sandbox --configuration release -Xswiftc -cross-module-optimization --enable-test-discovery
warning: '--enable-test-discovery' option is deprecated; tests are automatically discovered on all platforms
/home/taku0/opt/toolbox: warning: /nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
Fetching https://github.com/vapor/console-kit.git from cache
Fetching https://github.com/tanner0101/mustache.git from cache
Fetching https://github.com/jpsim/Yams.git from cache
Fetching https://github.com/apple/swift-nio.git from cache
Fetching https://github.com/apple/swift-log.git from cache
Cloning https://github.com/apple/swift-log.git
Resolving https://github.com/apple/swift-log.git at 1.4.2
Cloning https://github.com/jpsim/Yams.git
Resolving https://github.com/jpsim/Yams.git at 2.0.0
Cloning https://github.com/vapor/console-kit.git
Resolving https://github.com/vapor/console-kit.git at 4.2.6
Cloning https://github.com/tanner0101/mustache.git
Resolving https://github.com/tanner0101/mustache.git at 0.1.1
Cloning https://github.com/apple/swift-nio.git
Resolving https://github.com/apple/swift-nio.git at 2.31.1
/home/taku0/opt/toolbox/.build/checkouts/Yams: warning: /nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
/home/taku0/opt/toolbox/.build/checkouts/mustache: warning: /nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
/home/taku0/opt/toolbox/.build/checkouts/console-kit: warning: /nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
/home/taku0/opt/toolbox/.build/checkouts/swift-nio: warning: /nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
/home/taku0/opt/toolbox/.build/checkouts/swift-log: warning: /nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: unrecognized command-line option '-target'
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: unrecognized command-line option '-fmodule-name=CNIOWindows'
gcc: error: unrecognized command-line option '-fmodules-cache-path=/home/taku0/opt/toolbox/.build/x86_64-unknown-linux-gnu/release/ModuleCache'
gcc: error: unrecognized command-line option '-fmodule-name=CYaml'
gcc: error: unrecognized command-line option '-fmodule-name=CYaml'
gcc: error: unrecognized command-line option '-fmodule-name=CYaml'
gcc: error: unrecognized command-line option '-fmodules-cache-path=/home/taku0/opt/toolbox/.build/x86_64-unknown-linux-gnu/release/ModuleCache'
gcc: error: unrecognized command-line option '-fmodules-cache-path=/home/taku0/opt/toolbox/.build/x86_64-unknown-linux-gnu/release/ModuleCache'
gcc: error: unrecognized command-line option '-fmodules-cache-path=/home/taku0/opt/toolbox/.build/x86_64-unknown-linux-gnu/release/ModuleCache'
gcc: error: unrecognized command-line option '-fmodule-name=CYaml'
gcc: error: unrecognized command-line option '-fmodules-cache-path=/home/taku0/opt/toolbox/.build/x86_64-unknown-linux-gnu/release/ModuleCache'
gcc: error: unrecognized command-line option '-fmodule-name=CYaml'
gcc: error: unrecognized command-line option '-fmodules-cache-path=/home/taku0/opt/toolbox/.build/x86_64-unknown-linux-gnu/release/ModuleCache'
gcc: error: unrecognized command-line option '-fmodule-name=CNIOAtomics'
gcc: error: unrecognized command-line option '-fmodules-cache-path=/home/taku0/opt/toolbox/.build/x86_64-unknown-linux-gnu/release/ModuleCache'
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: unrecognized command-line option '-fmodule-name=CYaml'
[0/15] Compiling reader.c
Swift/ErrorType.swift:200: Fatal error: Error raised at top level: build.ShellError(terminationStatus: 1)
Current stack trace:
0    libswiftCore.so                    0x00007f268beaf410 swift_reportError + 50
1    libswiftCore.so                    0x00007f268bf24890 _swift_stdlib_reportFatalErrorInFile + 112
2    libswiftCore.so                    0x00007f268bc0ce40 _assertionFailure(_:_:file:line:flags:) + 1329
3    libswiftCore.so                    0x00007f268bc7c0a0 swift_errorInMain + 1075
4    build                              0x00005600ad0f643b <unavailable> + 9275
5    libc.so.6                          0x00007f268b69f6b0 __libc_start_main + 208
6    build                              0x00005600ad0f619a <unavailable> + 8602
make: *** [Makefile:5: build] Illegal instruction (core dumped)
make: *** Deleting file 'build'
```

</details>

**Comment by trepetti** [comment]
That looks like an issue building the C portion of the Yams library. You also need Clang to compile the native part of that. Do you have it installed in that shell?

**Comment by trepetti** [comment]
Or I guess it should be installed as part of Swift. Strange how it is defaulting to GCC. Will investigate.

**Comment by taku0** [comment]
GCC is used for some reason even if Clang is installed.  It produces slightly different error.

<details><summary>log</summary>

```
[nix-shell:~/.cache/nixpkgs-review/pr-132325-1]$ clang --version
clang version 10.0.0
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /nix/store/y8fc00i47y0fii3r4ij7sk9q1nrv0ll5-swift-5.4.2/bin

[nix-shell:~/.cache/nixpkgs-review/pr-132325-1]$ swift --version
Swift version 5.4.2 (swift-5.4.2-RELEASE)
Target: x86_64-unknown-linux-gnu

[nix-shell:~/.cache/nixpkgs-review/pr-132325-1]$ cd ~/opt/

[nix-shell:~/opt]$ git clone --depth 1 https://github.com/vapor/toolbox.git
Cloning into 'toolbox'...
remote: Enumerating objects: 52, done.
remote: Counting objects: 100% (52/52), done.
remote: Compressing objects: 100% (48/48), done.
remote: Total 52 (delta 1), reused 22 (delta 0), pack-reused 0
Receiving objects: 100% (52/52), 26.31 KiB | 13.15 MiB/s, done.
Resolving deltas: 100% (1/1), done.

[nix-shell:~/opt]$ cd toolbox/

[nix-shell:~/opt/toolbox]$ LC_ALL=C make
swiftc ./scripts/build.swift
./scripts/build.swift:48:10: warning: 'launchPath' is deprecated: renamed to 'executableURL'
    task.launchPath = "/usr/bin/env"
         ^
./scripts/build.swift:48:10: note: use 'executableURL' instead
    task.launchPath = "/usr/bin/env"
         ^~~~~~~~~~
         executableURL
./scripts/build.swift:50:10: warning: 'launch()' is deprecated: renamed to 'run'
    task.launch()
         ^
./scripts/build.swift:50:10: note: use 'run' instead
    task.launch()
         ^~~~~~
         run
./scripts/build.swift:61:10: warning: 'launchPath' is deprecated: renamed to 'executableURL'
    task.launchPath = "/usr/bin/env"
         ^
./scripts/build.swift:61:10: note: use 'executableURL' instead
    task.launchPath = "/usr/bin/env"
         ^~~~~~~~~~
         executableURL
./scripts/build.swift:69:10: warning: 'launch()' is deprecated: renamed to 'run'
    task.launch()
         ^
./scripts/build.swift:69:10: note: use 'run' instead
    task.launch()
         ^~~~~~
         run
/nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
./build
$ swift build --disable-sandbox --configuration release -Xswiftc -cross-module-optimization --enable-test-discovery
warning: '--enable-test-discovery' option is deprecated; tests are automatically discovered on all platforms
/home/taku0/opt/toolbox: warning: /nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
Fetching https://github.com/vapor/console-kit.git from cache
Fetching https://github.com/tanner0101/mustache.git from cache
Fetching https://github.com/jpsim/Yams.git from cache
Fetching https://github.com/apple/swift-nio.git from cache
Fetching https://github.com/apple/swift-log.git from cache
Cloning https://github.com/apple/swift-log.git
Resolving https://github.com/apple/swift-log.git at 1.4.2
Cloning https://github.com/apple/swift-nio.git
Resolving https://github.com/apple/swift-nio.git at 2.31.1
Cloning https://github.com/vapor/console-kit.git
Resolving https://github.com/vapor/console-kit.git at 4.2.6
Cloning https://github.com/tanner0101/mustache.git
Resolving https://github.com/tanner0101/mustache.git at 0.1.1
Cloning https://github.com/jpsim/Yams.git
Resolving https://github.com/jpsim/Yams.git at 2.0.0
/home/taku0/opt/toolbox/.build/checkouts/Yams: warning: /nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
/home/taku0/opt/toolbox/.build/checkouts/mustache: warning: /nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
/home/taku0/opt/toolbox/.build/checkouts/console-kit: warning: /nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
/home/taku0/opt/toolbox/.build/checkouts/swift-nio: warning: /nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
/home/taku0/opt/toolbox/.build/checkouts/swift-log: warning: /nix/store/hy3lz2vfv9qq2v5jz9nzlx6mmiaq79rj-binutils-2.35.1/bin/ld.gold: warning: /nix/store/cvr0kjg2q7z2wwhjblx6c73rv422k8cm-glibc-2.33-47/lib/../lib64/Scrt1.o: unknown program property type 0xc0008002 in .note.gnu.property section
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: x86_64-unknown-linux-gnu: No such file or directory
gcc: error: unrecognized command-line option '-target'
gcc: error: unrecognized command-line option '-fblocks'
gcc: error: unrecognized command-line option '-fmodules'; did you mean '-fmoduleinfo'?
gcc: error: unrecognized command-line option '-fmodule-name=CYaml'
gcc: error: unrecognized command-line option '-fmodules-cache-path=/home/taku0/opt/toolbox/.build/x86_64-unknown-linux-gnu/release/ModuleCache'
gcc: error: unrecognized command-line option '-fmodule-name=CYaml'
gcc: error: unrecognized command-line option '-fmodules-cache-path=/home/taku0/opt/toolbox/.build/x86_64-unknown-linux-gnu/release/ModuleCache'
gcc: error: unrecognized command-line option '-fmodule-name=CNIOLinux'
gcc: error: unrecognized command-line option '-fmodules-cache-path=/home/taku0/opt/toolbox/.build/x86_64-unknown-linux-gnu/release/ModuleCache'
gcc: error: unrecognized command-line option '-fmodule-name=CNIODarwin'
gcc: error: unrecognized command-line option '-fmodules-cache-path=/home/taku0/opt/toolbox/.build/x86_64-unknown-linux-gnu/release/ModuleCache'
[0/15] Compiling c-atomics.c
Swift/ErrorType.swift:200: Fatal error: Error raised at top level: build.ShellError(terminationStatus: 1)
Current stack trace:
0    libswiftCore.so                    0x00007fb1de3b3410 swift_reportError + 50
1    libswiftCore.so                    0x00007fb1de428890 _swift_stdlib_reportFatalErrorInFile + 112
2    libswiftCore.so                    0x00007fb1de110e40 _assertionFailure(_:_:file:line:flags:) + 1329
3    libswiftCore.so                    0x00007fb1de1800a0 swift_errorInMain + 1075
4    build                              0x0000560b190db4ab <unavailable> + 9387
5    libc.so.6                          0x00007fb1ddba36b0 __libc_start_main + 208
6    build                              0x0000560b190db20a <unavailable> + 8714
make: *** [Makefile:5: build] Illegal instruction (core dumped)
make: *** Deleting file 'build'
```

</details>

**Comment by trepetti** [comment]
Do you have `$CC` set as part of your shell environment by any chance? I was able to reproduce your issue by setting `$CC` to GCC. Apparently Swift will use that environment variable to assume you wanna use another version of Clang (but doe not support GCC). Unset that environment variable and give it a go.

**Comment by taku0** [comment]
Yes, `$CC` is `gcc` and without it, `make` succeeds.  I have no idea  why the variable is set.  Quick search on `~/.[a-z]*` with `CC=` don't matches interesting files.

**Comment by trepetti** [comment]
The `nixpkgs-review` shell is impure by default I think unless you set yourself as a trusted user in your nix configuration. The environment variable wasn't set on mine, so it seems to be externally to the project.

**Comment by trepetti** [comment]
Or I guess I've never figured out how to do that. The trusted user bit only controls sandboxing in the build, I guess. A pure shell for reviews would be a handy feature. Have had similar stuff crop up before.

**Comment by taku0** [comment]
Outside of `nixpkgs-review`, `$CC` is not set.  `nixpkgs-review pr 132325` somehow set the variable.

```
[taku0@nuc:~/src/nixpkgs]$ echo $CC


[taku0@nuc:~/src/nixpkgs]$ nix run nixpkgs.nixpkgs-review -c nixpkgs-review pr 132325
$ git -c fetch.prune=false fetch --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/132325/head:refs/nixpkgs-review/1
remote: Enumerating objects: 164, done.
remote: Counting objects: 100% (155/155), done.
remote: Compressing objects: 100% (65/65), done.
remote: Total 164 (delta 91), reused 139 (delta 87), pack-reused 9
Receiving objects: 100% (164/164), 1.52 MiB | 8.04 MiB/s, done.
Resolving deltas: 100% (91/91), completed with 46 local objects.
From https://github.com/NixOS/nixpkgs
   b5fa7e6c766..48bc8a78654  master     -> refs/nixpkgs-review/0
$ git worktree add /home/taku0/.cache/nixpkgs-review/pr-132325-2/nixpkgs 48bc8a78654d5be9a20eb64bdb54472377301f3c
Preparing worktree (detached HEAD 48bc8a78654)
Updating files: 100% (27360/27360), done.
HEAD is now at 48bc8a78654 Merge pull request #133137 from SuperSandro2000/fwts
$ git merge --no-commit --no-ff 5b8ef7ee01960266ecb04c507d0cfdbc1d2cfe30
Removing pkgs/development/compilers/swift/purity.patch
Removing pkgs/development/compilers/swift/patches/swift-package-manager-glibc-2.30.patch
Removing pkgs/development/compilers/swift/patches/swift-llvm.patch
Removing pkgs/development/compilers/swift/patches/llvm-toolchain-dir.patch
Removing pkgs/development/compilers/swift/patches/glibc-arch-headers.patch
Automatic merge went well; stopped before committing as requested
warning: ignoring the user-specified setting 'experimental-features', because it is a restricted setting and you are not a trusted user
warning: ignoring the user-specified setting 'system', because it is a restricted setting and you are not a trusted user
$ nix --experimental-features nix-command build --no-link --keep-going --option build-use-sandbox relaxed -f /home/taku0/.cache/nixpkgs-review/pr-132325-2/build.nix
warning: ignoring the user-specified setting 'experimental-features', because it is a restricted setting and you are not a trusted user
warning: ignoring the user-specified setting 'sandbox', because it is a restricted setting and you are not a trusted user

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/132325

1 package built:
swift

warning: ignoring the user-specified setting 'experimental-features', because it is a restricted setting and you are not a trusted user
$ nix-shell /home/taku0/.cache/nixpkgs-review/pr-132325-2/shell.nix

[nix-shell:~/.cache/nixpkgs-review/pr-132325-2]$ echo $CC
gcc

```

**Comment by taku0** [comment]
This may be a problem of `nixpkgs-review` because the variable is also set for other PRs. We can ignore this and proceed.

**Comment by taku0** [comment]
Okay, even `nix-shell --pure -p coreutils` sets various variables including `CC`, even in a container made with `docker run -it --rm nixos/nix`.  After all, `nix-shell --pure` reproduces an “environment that more closely corresponds to the ‘real’ Nix build”, rather than a shell of a fresh install of NixOS .

**Comment by trepetti** [comment]
Interesting. Thank you for your help digging into this.

**Comment by taku0** [comment]
This means that if we write a nix package that run `swift build` in `buildPhase`, it may result in the errors.  Maybe we should wrap `swift` (and `swiftc`?) with a wrapper that set `CC=${clang}/bin/clang`.  Anyway, it should be a separate pull request.

**Comment by trepetti** [comment]
@taku0 this last push wraps `swift` and `swiftc` with the correct `$CC`. I was able to confirm that the builds work in a shell too, which will help should we want to package software that requires Swift.

**Comment by trepetti** [comment]
Result of `nixpkgs-review pr 132325` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>swift</li>
  </ul>
</details>

**Comment by trepetti** [comment]
@SuperSandro2000 anything else you think we should modify on this one?

**Pull request merged**